Mathlib/Tactic/Simps/Basic.lean

```diff
@@ -637,55 +635,68 @@ def checkForUnusedCustomProjs (stx : Syntax) (str : Name) (projs : Array ParsedP
       m!"Not all of the custom declarations {customDeclarations} are used. Double check the {
         ""}spelling, and use `?` to get more information."
 
-/-- Data about default coercions. An entry consists of
-  `(projName, (className, functionName, arity))`, where
-  * `projName` is the name of a projection in a structure that must be used to trigger the search
-    for a coercion as projection.
-  * `className` is the name of the class we are looking for
-  * `functionName` is the function that we are using for a coercion as projection
-    (this is typically the first field of `className`).
-  * `arity` is the number of arguments of `className`. -/
-def defaultCoercions : NameMap (Name × Name × Nat) :=
-.ofList [(`toFun, (`FunLike, `FunLike.coe, 3)), (`carrier, (`SetLike, `SetLike.coe, 2))]
+/-- If a structure has a field that corresponds to a coercion to functions or sets, or corresponds
+  to notation, find the custom projection that uses this coercion or notation.
+  Returns the custom projection and the name of the projection used.
+
+We catch most errors this function causes, so that we don't fail if an unrelated projection has
+an applicable name. (e.g. `Iso.inv`)
+
+Implementation note: getting rid of TermElabM is tricky, since `Expr.mkAppOptM` doesn't allow to
+keep metavariables around, which are necessary for `OutParam`. -/
+def findAutomaticProjectionsAux (str : Name) (proj : ParsedProjectionData) (args : Array Expr) :
+  TermElabM <| Option (Expr × Name)  := do
+  if let some ⟨className, isNotation, findArgs⟩ :=
+    notationClassAttr.find? (← getEnv) proj.strName then
+    let findArgs ← unsafe evalConst findArgType findArgs
+    let classArgs ← try findArgs str className args
+    catch ex =>
+      trace[simps.debug] "Projection {proj.strName} is likely unrelated to the projection of {
+        className}:\n{ex.toMessageData}"
+      return none
+    let classArgs ← classArgs.mapM fun e => match e with
+      | none => mkFreshExprMVar none
+      | some e => pure e
+    let classArgs := classArgs.map Arg.expr
+    let projName := (getStructureFields (← getEnv) className)[0]!
+    let projName := className ++ projName
+    let eStr := mkAppN (← mkConstWithLevelParams str) args
+    let eInstType ←
+      try withoutErrToSorry (elabAppArgs (← Term.mkConst className) #[] classArgs none true false)
+      catch ex =>
+        trace[simps.debug] "Projection doesn't have the right type for the automatic projection:\n{
+          ex.toMessageData}"
+        return none
```

Comment:
It seems incredibly unlikely, since the classes that will be used here are notation classes like `Mul` and `HasSubset`, and those shouldn't have other type-class arguments. And if there are, I probably need to change more, because I made all meta variables using `mkFreshExprMVar`, and I probably need to make some of them synthetic, if I wanted to support this case (which I probably don't want to).